File: compute/backend/routes/logs.py
from flask import Blueprint, Response, request
from objects.user import User
from dataclasses import asdict
import pandas as pd

"""
Logs Blueprint (Flask)

Provides HTTP endpoints for managing habit log entries.

INTEGRATION NOTES FOR HABO (Flutter) FRONTEND
-----------------------------------------------
The current `/logs` routes are designed for a simpler log model (state strings,
self_reported flags) and do NOT match Habo's event system.

Habo uses a different concept called "Events" (see `mobile/Habo/lib/repositories/event_repository.dart`):
  - Events are tied to a specific habit ID and a specific date (DateTime).
  - Each event is a List containing:
    - [0]: DayType enum (clear, check, fail, skip, progress)
    - [1]: comment (string, optional)
    - [2]: progressValue (double, optional, only for numeric habits with DayType.progress)

The mobile app's `HttpEventRepository` expects routes under `/<user_id>/events`:
  - POST   /<user_id>/events/add          (JSON body: {habitId, date, eventData})
  - POST   /<user_id>/events/delete        (JSON body: {habitId, date})
  - GET    /<user_id>/events/habit/<habit_id>
  - GET    /<user_id>/events/habit/<habit_id>/map
  - POST   /<user_id>/events/habit/<habit_id>/clear
  - POST   /<user_id>/events/habit/<habit_id>/batch  (JSON body: {habitId, events: {date: eventData}})
  - POST   /<user_id>/events/clear-all

See `backend/HABO_API_SPECIFICATION.md` for the full event API specification.

TODO (backend): Decide whether to:
  1. Extend these `/logs` routes to also support Habo's event format, OR
  2. Create new `/events` routes specifically for Habo and keep `/logs` for your
     existing IoT/device logging use case.
"""

logs_bp = Blueprint("logs", __name__)

@logs_bp.post(f'/<int:user_id>/{logs_bp.name}/add')
def post_add_log(user_id: int):
    user = User(user_id)
    
    habit_name: str = request.args.get('name')
    state: str = request.args.get('state')
    self_reported: bool = request.args.get('reported', type=bool)
    
    user.log_habit(habit_name, state, self_reported)
    
    return {
        "result" : f"Successfully logged habit {habit_name} ({state}:{self_reported})"
    }, 200

@logs_bp.get(f'/<int:user_id>/{logs_bp.name}')
def get_logs(user_id: int):
    user = User(user_id)
    
    habit_name: str = request.args.get('name')
    
    df: pd.DataFrame = user.get_habit_logs(habit_name)
    
    if df is None:
        df = pd.DataFrame()
    
    return {
        "result" : df.to_dict(orient="records")
    }, 200

@logs_bp.post(f'/<int:user_id>/{logs_bp.name}/update')
def post_update_log(user_id: int):
    user = User(user_id)
    
    id: int = request.args.get('id', type=int)
    field: str = request.args.get('field')
    value: str = request.args.get('value')
    
    # TODO: need to verify field value change in type, could break here.
    
    user.update_log(id, field, value)
    
    return {
        "result" : f"Successfully updated log {id} {field} -> {value}"
    }, 200

@logs_bp.post(f'/<int:user_id>/{logs_bp.name}/delete')
def post_delete_log(user_id: int):
    user = User(user_id)
    
    id: int = request.args.get('id', type=int)
    
    user.delete_log(id)
    
    return {
        "result" : f"Successfully deleted log {id}"
    }, 200

File: compute/mobile/logs.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

const String baseUrl = 'http://192.168.2.19:5000';

/// Adds a new log entry for the user.
///
/// Makes a POST request to /<user_id>/logs/add with query parameters.
/// @param userId The user ID
/// @param name The habit name
/// @param state The log state
/// @param reported Whether self-reported
/// @return True if successful
Future<bool> addLog(int userId, String name, String state, bool reported) async {
  final url = '$baseUrl/$userId/logs/add';
  final uri = Uri.parse(url).replace(queryParameters: {
    'name': name,
    'state': state,
    'reported': reported.toString(),
  });

  final response = await http.post(uri);

  return response.statusCode == 200;
}

/// Retrieves log entries for the user, optionally filtered by habit name.
///
/// Makes a GET request to /<user_id>/logs.
/// @param userId The user ID
/// @param name Optional habit name to filter
/// @return List of log maps
Future<List<Map<String, dynamic>>> getLogs(int userId, {String? name}) async {
  final url = '$baseUrl/$userId/logs';
  final params = name != null ? {'name': name} : {};
  final uri = Uri.parse(url).replace(queryParameters: params);

  final response = await http.get(uri);

  if (response.statusCode == 200) {
    final data = jsonDecode(response.body) as Map<String, dynamic>;
    final result = data['result'] as List<dynamic>;
    return result.cast<Map<String, dynamic>>();
  } else {
    throw Exception('Failed to load logs: ${response.statusCode}');
  }
}

/// Updates a log entry for the user.
///
/// Makes a POST request to /<user_id>/logs/update.
/// @param userId The user ID
/// @param id The log ID
/// @param field The field to update
/// @param value The new value
/// @return True if successful
Future<bool> updateLog(int userId, int id, String field, String value) async {
  final url = '$baseUrl/$userId/logs/update';
  final uri = Uri.parse(url).replace(queryParameters: {
    'id': id.toString(),
    'field': field,
    'value': value,
  });

  final response = await http.post(uri);

  return response.statusCode == 200;
}

/// Deletes a log entry for the user.
///
/// Makes a POST request to /<user_id>/logs/delete.
/// @param userId The user ID
/// @param id The log ID
/// @return True if successful
Future<bool> deleteLog(int userId, int id) async {
  final url = '$baseUrl/$userId/logs/delete';
  final uri = Uri.parse(url).replace(queryParameters: {
    'id': id.toString(),
  });

  final response = await http.post(uri);

  return response.statusCode == 200;
}
